File: Data Structures & Algorithms/three-integer-sum/submission-6.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        print(nums)
        for c, i in enumerate(nums):
            if c > 0 and nums[c-1] == nums[c]:
                continue
            
            l = c + 1
            r = len(nums) - 1
            while l < r:
                target = i + nums[l] + nums[r]
                if target > 0:
                    r -= 1
                elif target < 0:
                    l += 1
                else:
                    res.append([i, nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-6.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];

        int prefix = 1;
        for (int i = 0; i < n; i++){
            res[i] = prefix;
            prefix *= nums[i];
        }

        int suffix = 1;
        for (int i = n - 1; i > -1; i--){
            res[i] *= suffix;
            suffix *= nums[i];
        }

        return res;
    }
}